diff --git a/Makefile.am b/Makefile.am
index f2899cc..979e47b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,7 @@
 ACLOCAL_AMFLAGS = -I m4
 
 AM_CPPFLAGS = $(all_includes) -I$(top_srcdir)/include
-SUBDIRS = include src src/vty src/codec src/gsm src/gb src/ctrl tests utils
+SUBDIRS = include src src/vty src/codec src/gsm src/gb tests utils
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libosmocore.pc libosmocodec.pc libosmovty.pc libosmogsm.pc \
diff --git a/src/macaddr.c b/src/macaddr.c
index 7bc4401..5fa6b4f 100644
--- a/src/macaddr.c
+++ b/src/macaddr.c
@@ -47,7 +47,7 @@ int osmo_macaddr_parse(uint8_t *out, const char *in)
 	return 0;
 }
 
-#if defined(__FreeBSD__)
+#if defined(__FreeBSD__) || (defined(__APPLE__) && defined(__MACH__))
 #include <sys/socket.h>
 #include <sys/types.h>
 #include <ifaddrs.h>
diff --git a/src/utils.c b/src/utils.c
index 6ece1a8..8dfe3ef 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -219,8 +219,8 @@ char *osmo_hexdump_nospc(const unsigned char *buf, int len)
 }
 
 	/* Compat with previous typo to preserve abi */
-char *osmo_osmo_hexdump_nospc(const unsigned char *buf, int len)
-	__attribute__((weak, alias("osmo_hexdump_nospc")));
+//char *osmo_osmo_hexdump_nospc(const unsigned char *buf, int len)
+//	__attribute__((weak, alias("osmo_hexdump_nospc")));
 
 #include "../config.h"
 #ifdef HAVE_CTYPE_H
